### PR TITLE
Graph rewrite

### PR DESCRIPTION
**Pull request by dboyliao**
`uTensorGraph` enhancement for graph rewriting

### CODE REVIEW

File: utensor_cgen/ir/base.py [hunk 1]
```diff
@@ -105,15 +129,14 @@ def check(self, attribute, value):
...
-  # ugraph = attr.ib(default=None, init=False)
+
   @property
   def input_nodes(self):
     in_ops = []
     for tensor in self.input_tensors:
       if tensor.op_name not in in_ops:
         in_ops.append(tensor.op_name)
-    return [self.ugraph.ops_info[name] for name in in_ops]
+    return [self.ugraph.ops_info.get(name, None) for name in in_ops]
```
Comment (mbartling) [inline utensor_cgen/ir/base.py]: A little confused, why did you replace the dict style lookup with get?

File: utensor_cgen/ir/base.py [hunk 2]
```diff
@@ -208,11 +248,24 @@ def graph_def(self):
...
-  @staticmethod
-  def _topologic_order_graph(ugraph):
+
+  def add_op(self, op):
+    if not isinstance(op, OperationInfo):
+      raise ValueError('expecting OperationInfo, get {}'.format(type(op)))
+    if op.name in self.ops_info:
+      raise ValueError('duplicate op detected, {}'.format(op.name))
+    self.ops_info[op.name] = op
+    self._topologic_order_graph()
```
Comment (mbartling) [inline utensor_cgen/ir/base.py]: So do we reorder after every add_op?

Comment (neil-tan) [reply utensor_cgen/ir/base.py]: yes

**Comment by neil-tan** [comment]
@mbartling This PR includes many support functions required by many graph-rewriting functions, for example, the isomorphic matcher.
I've just rebased cmsis_nn branch to this PR. Stuff like this here and there: https://github.com/uTensor/utensor_cgen/blob/b0989580d4a51c7b61ab240d914d2c7fead4f378/utensor_cgen/transformer/cmsis_nn.py#L413-L414

**Pull request merged**